src/main/java/bio/terra/service/snapshot/flight/duos/SnapshotUpdateDuosDatasetFlight.java

```diff
@@ -0,0 +1,58 @@
+package bio.terra.service.snapshot.flight.duos;
+
+import bio.terra.common.iam.AuthenticatedUserRequest;
+import bio.terra.model.DuosFirecloudGroupModel;
+import bio.terra.service.auth.iam.IamService;
+import bio.terra.service.duos.DuosDao;
+import bio.terra.service.duos.DuosService;
+import bio.terra.service.job.JobMapKeys;
+import bio.terra.service.snapshot.SnapshotDao;
+import bio.terra.stairway.Flight;
+import bio.terra.stairway.FlightMap;
+import java.util.UUID;
+import org.springframework.context.ApplicationContext;
+
+/**
+ * This flight links a snapshot to a DUOS dataset for the purposes of syncing its authorized users
+ * as snapshot readers via a TDR-managed Firecloud group. It supports linking, unlinking, and
+ * updating a snapshot's existing link (unlinking + linking).
+ */
+public class SnapshotUpdateDuosDatasetFlight extends Flight {
+
+  public SnapshotUpdateDuosDatasetFlight(FlightMap inputParameters, Object applicationContext) {
+    super(inputParameters, applicationContext);
+
+    ApplicationContext appContext = (ApplicationContext) applicationContext;
+    DuosDao duosDao = appContext.getBean(DuosDao.class);
+    DuosService duosService = appContext.getBean(DuosService.class);
+    IamService iamService = appContext.getBean(IamService.class);
+    SnapshotDao snapshotDao = appContext.getBean(SnapshotDao.class);
+
+    AuthenticatedUserRequest userReq =
+        inputParameters.get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    UUID snapshotId = inputParameters.get(JobMapKeys.SNAPSHOT_ID.getKeyName(), UUID.class);
+    String duosId = inputParameters.get(SnapshotDuosMapKeys.DUOS_ID, String.class);
+    DuosFirecloudGroupModel firecloudGroupPrev =
+        inputParameters.get(
+            SnapshotDuosMapKeys.FIRECLOUD_GROUP_PREV, DuosFirecloudGroupModel.class);
+
+    boolean linking = (duosId != null);
+    boolean unlinking = (firecloudGroupPrev != null);
+
+    if (unlinking) {
```

Comment:
Thoughts on using optional steps here instead of this branching logic?

Reply:
I suppose that would mean there would be nested optional steps with the IfNoGroupRetrievedStep optional steps below.

Reply:
Ah yes, this is a good question!

I felt like this would significantly detract from readability, making nearly every step optional (or an optional-of-an-optional).  I want to say that we lose some clarity in logging when we use optional steps.  For the Firecloud group creation steps, I favored the optional steps since we're checking the condition against the working map.  But… I don't love the inconsistency.

When looking at existing flights, I saw that we handle conditional step running a few different ways:
- Via optional steps
- Via flight-level conditionals to determine whether or not to run the step (as I did here)
- Via step-level conditionals - the step itself is run but maybe boils down to a no-op

I liked the logical definition of distinguishing between steps that unlink vs. steps that link, and since there would be cases where we wanted to unlink and link I wanted to handle it all in one flight.

Reply:
I wonder if there's an opportunity to expand our flight logging hook to expose an optional's inner step, similar to what I did for the flight construction unit test?  I can timebox myself and look.

https://github.com/DataBiosphere/jade-data-repo/blob/9220aca71e5c84d8f9d0619ec7012af56c76197c/src/test/java/bio/terra/service/snapshot/flight/duos/SnapshotUpdateDuosDatasetFlightTest.java#L110-L115

Reply:
Yeah, I actually wonder if it would make sense to create a base step class for ourselves and all of our steps could extend that class.  That common class would have an overridable "get my real class" method that could be implemented (there are a number of step classes that suffer from this issue, not just optional)

Reply:
Interesting idea re: a common base class.  I'd be interested in talking more about the other step classes which could benefit from such an override!

I did make a little progress poking around the `OptionalStep` as well as `StairwayLoggingHooks` but I'll table it for now / write a separate ticket for further investigation.